<?xml version="1.0"?>
<launch>
  <arg name="use_sim_time" default="true"/>
  <arg name="livox_scan_topic" default="/livox_scan"/>
  <arg name="livox_points_topic" default="/livox_points"/>
  <arg name="xacro_file_path" default="$(find-pkg-share orange_description)/xacro/orange_robot.xacro"/>

  <!-- gyrobias corrector
  <node pkg="cxd5602pwbimu" exec="gyrobias_corrector">
    <param name="use_sim_time" value="$(var use_sim_time)"/>
  </node> -->
  <!-- madgwick filter node
  <node pkg="imu_filter_madgwick" exec="imu_filter_madgwick_node" name="madgwick_filter">
    <param name="use_sim_time" value="$(var use_sim_time)"/>
    <param from="$(find-pkg-share cxd5602pwbimu)/config/madgwick.yaml"/>
    <remap from="/imu/data_raw" to="/imu/spresense"/>
    <remap from="/imu/data" to="/imu/spresense/madgwick"/>
  </node> -->
  <!--initial yaw corrector
  <node pkg="cxd5602pwbimu" exec="initial_yaw_corrector" name="spimu_yaw_corrector">
    <param name="use_sim_time" value="$(var use_sim_time)"/>
    <param name="input_topic" value="/imu/spresense/madgwick"/>
    <param name="output_topic" value="/imu/spresense/inityaw2"/>
  </node> -->
  <!-- wheel_imu_odom for spresense imu
  <node pkg="cxd5602pwbimu" exec="wheel_spimu_odom" name="madg_spimu_odom">
    <param name="use_sim_time" value="$(var use_sim_time)"/>
    <param name="odom_topic" value="/odom"/>
    <param name="imu_topic" value="/imu/spresense/inityaw2"/>
    <param name="fused_odom_topic" value="/odom/wheel_spimu2"/>
    <param name="publish_TF" value="false"/>
  </node> -->

  <!-- quat attitude estimator -->
  <node pkg="cxd5602pwbimu" exec="quat_attitude_estimator">
    <param name="use_sim_time" value="$(var use_sim_time)"/>
    <param name="input_topic" value="/imu/spresense"/>
    <param name="output_topic" value="/imu/spresense/inityaw"/>
  </node>
  <!-- wheel_imu_odom for spresense imu -->
  <node pkg="cxd5602pwbimu" exec="wheel_spimu_odom" name="quat_spimu_odom">
    <param name="use_sim_time" value="$(var use_sim_time)"/>
    <param name="odom_topic" value="/odom"/>
    <param name="imu_topic" value="/imu/spresense/inityaw"/>
    <param name="fused_odom_topic" value="/odom/wheel_spimu"/>
    <param name="publish_TF" value="false"/>
  </node>

  <!-- initial yaw corrector
  <node pkg="cxd5602pwbimu" exec="initial_yaw_corrector" name="imu_yaw_corrector">
    <param name="use_sim_time" value="$(var use_sim_time)"/>
    <param name="input_topic" value="/imu"/>
    <param name="output_topic" value="/imu/inityaw"/>
  </node> -->
  <!-- wheel_imu_odom for imu
  <node pkg="orange_sensor_tools" exec="wheel_imu_odom">
    <param name="use_sim_time" value="$(var use_sim_time)"/>
    <param name="odom_topic" value="/odom"/>
    <param name="imu_topic" value="/imu/inityaw"/>
    <param name="publish_TF" value="false"/>
  </node> -->

  <!-- GPSposition converter
  <node pkg="orange_gnss" exec="lonlat_to_odom" output="screen">
    <param name="use_sim_time" value="$(var use_sim_time)"/>
    <param name="Position_magnification" value="1.675"/>
  </node> -->
  <!-- combination_GPSposition_GPSheading
  <node pkg="orange_gnss" exec="combination" output="screen">
    <param name="use_sim_time" value="$(var use_sim_time)"/>
  </node> -->
  <!--initial xy corrector -->
  <node pkg="cxd5602pwbimu" exec="initial_xy_corrector">
    <param name="use_sim_time" value="$(var use_sim_time)"/>
    <param name="input_topic" value="/odom/UM982"/>
    <param name="output_topic" value="/odom/UM982/initxy"/>
  </node>

  <!-- ekf_myself -->
  <!--<node pkg="orange_gnss" exec="ekf_myself" output="screen">
    <param name="use_sim_time" value="$(var use_sim_time)"/>
    <param name="ekf_publish_TF" value="false"/>
  </node> -->
  <node pkg="try_navigation" exec="ekf_myself_gps" output="screen">
    <param name="use_sim_time" value="$(var use_sim_time)"/>
    <param name="ekf_publish_TF" value="false"/>
  </node>
  <node pkg="sensor_utils" exec="yaw_kf_node.py">
    <param name="use_sim_time" value="$(var use_sim_time)"/>
  </node>
  <node pkg="sensor_utils" exec="sat_kf_node.py">
    <param name="use_sim_time" value="$(var use_sim_time)"/>
  </node>
  <node pkg="sensor_utils" exec="vec_kf_node.py">
    <param name="use_sim_time" value="$(var use_sim_time)"/>
    <param name="publish_TF" value="false"/>
  </node>
  <node pkg="sensor_utils" exec="nis_kf_node.py">
    <param name="use_sim_time" value="$(var use_sim_time)"/>
  </node>
  <node pkg="sensor_utils" exec="my_kf_node.py">
    <param name="use_sim_time" value="$(var use_sim_time)"/>
    <param name="publish_TF" value="false"/>
  </node>

  <node pkg="sensor_utils" exec="my2_kf_node.py">
    <param name="use_sim_time" value="$(var use_sim_time)"/>
  </node>

  <!-- fast lio -->
  <include file="$(find-pkg-share fast_lio)/launch/mapping.launch.py">
    <arg name="use_sim_time" value="$(var use_sim_time)"/>
  </include>
  <node pkg="fast_odom_convert" exec="fast_odom_convert">
    <param name="use_sim_time" value="$(var use_sim_time)"/>
  </node>

  <!-- livox_to_pointcloud2 -->
  <node pkg="livox_to_pointcloud2" exec="livox_to_pointcloud2_node">
    <param name="use_sim_time" value="$(var use_sim_time)"/>
    <remap from="/livox_pointcloud" to="/livox/lidar"/>
  </node>
  <!-- ground_segmentation2
  <include file="$(find-pkg-share pcd_convert)/launch/pcd_convert.launch.py">
    <arg name="use_sim_time" value="$(var use_sim_time)"/>
  </include> -->
  <!-- pointcloud2_to_laserscan
  <include file="$(find-pkg-share orange_sensor_tools)/launch/livox_to_pointcloud2_laserscan.launch.py">
    <arg name="use_sim_time" value="$(var use_sim_time)"/>
    <arg name="cloud_in" value="/pcd_segment_obs"/>
    <arg name="scan_out" value="$(var livox_scan_topic)"/>
  </include> -->
  <!-- slam
  <include file="$(find-pkg-share orange_slam)/launch/slam_toolbox.launch.xml">
    <arg name="use_sim_time" value="$(var use_sim_time)"/>
  </include> -->

  <!-- robot_state_publisher -->
  <node pkg="robot_state_publisher" exec="robot_state_publisher">
    <param name="use_sim_time" value="$(var use_sim_time)"/>
    <param name="robot_description" value="$(command 'xacro $(var xacro_file_path)')"/>
  </node>
  <!-- joint_state_publisher -->
  <node pkg="joint_state_publisher" exec="joint_state_publisher">
    <param name="use_sim_time" value="$(var use_sim_time)"/>
    <param name="robot_description" value="$(command 'xacro $(var xacro_file_path)')"/>
  </node>
</launch>
